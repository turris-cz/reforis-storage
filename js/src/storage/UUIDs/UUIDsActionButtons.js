/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";

import { Button } from "foris";

UUIDsActionButtons.propTypes = {
    onUnselectSrv: PropTypes.func.isRequired,
    onUpdateSrv: PropTypes.func.isRequired,
    storageIsPending: PropTypes.bool.isRequired,
};

export default function UUIDsActionButtons({ onUnselectSrv, onUpdateSrv, storageIsPending }) {
    return (
        <>
            <Button
                className="btn-primary offset-lg-1 col-lg-4 col-sm-12"
                disabled={storageIsPending}
                onClick={onUnselectSrv}
            >
                {_("Unset UUID")}
            </Button>
            <Button
                className="btn-primary col-sm-12 col-lg-4 offset-lg-2 col-lg-3"
                disabled={storageIsPending}
                onClick={onUpdateSrv}
            >
                {_("Set UUID")}
            </Button>
        </>
    );
}
