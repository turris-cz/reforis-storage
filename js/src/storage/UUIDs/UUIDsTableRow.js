import React from "react";

export default function UUIDsTableRow({
    uuid, drives, selected, setSelectedUUID, disabled,
}) {
    const devices = drives.map((drive) => drive.dev).join(", ");

    return (
        <tr className={selected ? "table-active" : ""}>
            <td>
                <div className="custom-control custom-radio">
                    <input
                        checked={selected}
                        type="radio"
                        id={uuid}
                        value={uuid}
                        name="uuids"
                        className="custom-control-input"
                        onChange={() => {
                            setSelectedUUID(uuid);
                        }}
                        disabled={disabled}
                    />
                    <label className="custom-control-label" htmlFor={uuid}>
                        {devices}
                    </label>
                </div>
            </td>
            <td>
                {uuid}
            </td>
        </tr>
    );
}
