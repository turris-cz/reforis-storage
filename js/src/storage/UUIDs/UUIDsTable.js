/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import PropTypes from "prop-types";

import UUIDsTableRow from "./UUIDsTableRow";

UUIDsTable.propTypes = {
    drivesByUUIDs: PropTypes.object.isRequired,
    selectedUUID: PropTypes.string.isRequired,
    setSelectedUUID: PropTypes.func.isRequired,
    storageIsPending: PropTypes.bool.isRequired,
};

export default function UUIDsTable({
    drivesByUUIDs,
    selectedUUID,
    setSelectedUUID,
    storageIsPending,
}) {
    return (
        <div className="table-responsive">
            <table className="table table-hover">
                <thead className="thead-light">
                    <tr>
                        <th scope="col">{_("Devices")}</th>
                        <th scope="col">{_("UUID")}</th>
                    </tr>
                </thead>
                <tbody>
                    {Object.keys(drivesByUUIDs).map((uuid) => (
                        <UUIDsTableRow
                            key={uuid}
                            uuid={uuid}
                            drives={drivesByUUIDs[uuid]}
                            // uuid can be empty, it means nothing is selected.
                            selected={uuid && uuid === selectedUUID}
                            setSelectedUUID={setSelectedUUID}
                            disabled={storageIsPending}
                        />
                    ))}
                </tbody>
            </table>
        </div>
    );
}
