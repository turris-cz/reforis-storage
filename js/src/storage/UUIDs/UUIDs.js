/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import { API_STATE, useAlert, useAPIPost } from "foris";

import API_URLs from "../../API";
import { filterNonBTRFS, groupDrivesByUUIDs } from "../utils";
import UUIDsTable from "./UUIDsTable";
import UUIDsActionButtons from "./UUIDsActionButtons";

UUIDs.propTypes = {
    drives: PropTypes.arrayOf(
        PropTypes.shape({
            dev: PropTypes.string.isRequired,
            description: PropTypes.string.isRequired,
            fs: PropTypes.string.isRequired,
            uuid: PropTypes.string.isRequired,
        })
    ).isRequired,
    currentUUID: PropTypes.string.isRequired,
    storageIsPending: PropTypes.bool.isRequired,
    updateUUIDCallback: PropTypes.func.isRequired,
};

UUIDs.defaultProps = {
    storageIsPending: false,
};

export default function UUIDs({
    drives,
    currentUUID,
    storageIsPending,
    updateUUIDCallback,
}) {
    const [selectedUUID, setSelectedUUID] = useState(currentUUID);
    const [postUpdateSrvStatus, postUpdateSrv] = useAPIPost(API_URLs.updateSrv);
    const [setAlert] = useAlert();
    useEffect(() => {
        if (postUpdateSrvStatus.state === API_STATE.ERROR) {
            setAlert(_("UUID selection failed."));
        } else if (postUpdateSrvStatus.state === API_STATE.SUCCESS) {
            updateUUIDCallback();
        }
    }, [postUpdateSrvStatus.state, setAlert, updateUUIDCallback]);

    function onUnselectSrv() {
        postUpdateSrv({ data: { uuid: "" } });
    }

    function onUpdateSrv() {
        postUpdateSrv({ data: { uuid: selectedUUID } });
    }

    const drivesByUUIDs = groupDrivesByUUIDs(filterNonBTRFS(drives));

    if (Object.keys(drivesByUUIDs).length === 0) {
        return (
            <p className="text-muted text-center">
                {_("There aren't prepared drives to be used.")}
            </p>
        );
    }

    const buttonIsDisabled =
        selectedUUID.length === 0 ||
        selectedUUID.length < 2 ||
        storageIsPending;

    return (
        <>
            <UUIDsTable
                drivesByUUIDs={drivesByUUIDs}
                selectedUUID={selectedUUID}
                setSelectedUUID={setSelectedUUID}
                storageIsPending={storageIsPending}
            />
            <UUIDsActionButtons
                onUnselectSrv={onUnselectSrv}
                onUpdateSrv={onUpdateSrv}
                buttonIsDisabled={buttonIsDisabled}
            />
        </>
    );
}
