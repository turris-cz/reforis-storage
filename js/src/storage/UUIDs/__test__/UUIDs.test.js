/*
 * Copyright (C) 2019-2021 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import { mockSetAlert } from "foris/testUtils/alertContextMock";
import {
    render,
    wait,
    fireEvent,
    getByLabelText,
    getByTestId,
} from "foris/testUtils/customTestRender";
import { mockJSONError } from "foris/testUtils/network";
import mockAxios from "jest-mock-axios";

import drives from "../../__tests__/__fixtures__/drives";
import UUIDs from "../UUIDs";

describe("<UUIDs />", () => {
    let container;
    const updateUUIDCallback = jest.fn();
    beforeEach(() => {
        ({ container } = render(
            <UUIDs
                currentUUID=""
                storageIsPending={false}
                updateUUIDCallback={updateUUIDCallback}
                {...drives}
            />
        ));
    });

    it("Should render.", () => {
        expect(container).toMatchSnapshot();
    });

    it("Should post on 'Use drive' button click.", () => {
        fireEvent.click(getByLabelText(container, "sdc"));
        fireEvent.click(getByTestId(container, "use-drive-btn"));

        const data = { uuid: "4444" };
        expect(mockAxios.post).toHaveBeenCalledWith(
            "/reforis/storage/api/update-srv",
            data,
            expect.anything()
        );
    });

    it("Should post on 'Disable external storage' button click.", async () => {
        fireEvent.click(getByLabelText(container, "sdc"));
        fireEvent.click(getByTestId(container, "disable-external-storage-btn"));

        const data = { uuid: "" };
        expect(mockAxios.post).toHaveBeenCalledWith(
            "/reforis/storage/api/update-srv",
            data,
            expect.anything()
        );
    });

    it("Should call updateUUIDCallback when 'Disable external storage' button click.", async () => {
        expect(updateUUIDCallback).not.toBeCalled();
        fireEvent.click(getByLabelText(container, "sdc"));
        fireEvent.click(getByTestId(container, "disable-external-storage-btn"));
        mockAxios.mockResponse({ result: true });
        await wait(() => expect(updateUUIDCallback).toBeCalled());
    });

    it("Should call updateUUIDCallback when 'Use drive' button click.", async () => {
        expect(updateUUIDCallback).not.toBeCalled();
        fireEvent.click(getByLabelText(container, "sdc"));
        fireEvent.click(getByTestId(container, "use-drive-btn"));
        mockAxios.mockResponse({ result: true });
        await wait(() => expect(updateUUIDCallback).toBeCalled());
    });

    it("Should handle Use drive post error.", async () => {
        fireEvent.click(getByLabelText(container, "sdc"));
        fireEvent.click(getByTestId(container, "use-drive-btn"));
        mockJSONError();
        await wait(() =>
            expect(mockSetAlert).toBeCalledWith("Drive selection failed.")
        );
    });

    it("Should handle Disable external storage post error.", async () => {
        fireEvent.click(getByLabelText(container, "sdc"));
        fireEvent.click(getByTestId(container, "disable-external-storage-btn"));
        mockJSONError();
        await wait(() =>
            expect(mockSetAlert).toBeCalledWith("Drive selection failed.")
        );
    });
});
