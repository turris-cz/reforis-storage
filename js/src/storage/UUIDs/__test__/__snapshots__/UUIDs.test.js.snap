// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UUIDs /> Should render. 1`] = `
<div>
  <table
    class="table table-hover col-sm-12 offset-lg-1 col-lg-10"
  >
    <thead>
      <tr>
        <th
          scope="col"
        >
          Devices
        </th>
        <th
          scope="col"
        >
          UUID
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class=""
      >
        <td>
          <div
            class="custom-control custom-radio custom-control-inline"
          >
            <input
              class="custom-control-input"
              id="4444"
              name="uuids"
              type="radio"
              value="4444"
            />
            <label
              class="custom-control-label"
              for="4444"
            >
              sdc
            </label>
          </div>
        </td>
        <td>
          4444
        </td>
      </tr>
    </tbody>
  </table>
  <button
    class="btn btn-primary offset-lg-1 col-lg-4 col-sm-12"
    type="button"
  >
     
     
    Unset UUID
  </button>
  <button
    class="btn btn-primary col-sm-12 col-lg-4 offset-lg-2 col-lg-3"
    type="button"
  >
     
     
    Set UUID
  </button>
</div>
`;
